src/dbnode/persist/fs/retriever.go

```diff
@@ -561,7 +573,10 @@ func (req *retrieveRequest) onCallerOrRetrieverDone() {
 
 func (req *retrieveRequest) Reset(segment ts.Segment) {
 	req.reader.Reset(segment)
-	req.resultWg.Done()
+	if req.err == nil {
+		// If there was an error, we've already called done.
```

Comment:
Yeah so I didn't change how these methods are called, its just that a bug in my code revealed that in certain error scenarios we would end up with a negative waitgroup. I looked through the code and it wasn't immediately obvious to me what the issue was. I can probably try and look through the code and see exactly how it could happen and fix it? I figured these guards would be good either way

Reply:
Actually I'm looking through the code now, are you sure this used to be correct?

We have this bit of code:

```golang
for _, req := range reqs {
		entry, err := seeker.SeekIndexEntry(req.id, seekerResources)
		if err != nil && err != errSeekIDNotFound {
			req.onError(err)
			continue
		}

		if err == errSeekIDNotFound {
			req.notFound = true
		}
		req.indexEntry = entry
	}
```

and later on in the function we do this to each `req` in a loop:

```golang
// Complete request.
req.onRetrieved(seg)
```

The only way `req.onRetrieved(seg)` won't be called is if:

```golang
// Only try to seek the ID if it exists, otherwise we'll get a checksum
// mismatch error because default offset value for indexEntry is zero.
if !req.notFound {
    data, err = seeker.SeekByIndexEntry(req.indexEntry, seekerResources)
        if err != nil && err != errSeekIDNotFound {
            req.onError(err)
            continue
        }
}
```

however `req.notFound = true` doesn't get set in the error case. So I guess we can fix it by changing the code to:

```golang
// Only try to seek the ID if it exists and there haven't been any errors already, otherwise we'll get a checksum mismatch error because default offset value for indexEntry is zero.
if !req.notFound {
    data, err = seeker.SeekByIndexEntry(req.indexEntry, seekerResources)
        if err != nil && err != errSeekIDNotFound {
            req.onError(err)
            continue
        }
}
```

Reply:
I ended up adding a variable: `foundAndHasNoError = !req.notFound && req.err == nil` and tweaking the code slightly, let me know what you think. Might be easiest to discuss in person

Reply:
Ok forget all that, after realizing how buggy the existing implementation was I added some unit tests and just generally refactored it to be correct. Let me know what you think

Reply:
Nice, yeah good call - this looks good now. Thanks for adding the tests too, nice work.